bootstrap/pkg/kfapp/gcp/gcp.go

```diff
@@ -477,7 +440,87 @@ func (gcp *Gcp) Apply(resources kftypes.ResourceEnum) error {
 	return nil
 }
 
+// Try to get information for the deployment. If returned, delete it.
+func deleteDeployment(deploymentmanagerService *deploymentmanager.Service, ctx context.Context,
+	project string, name string) error {
+	_, err := deploymentmanagerService.Deployments.Get(project, name).Context(ctx).Do()
+	if err != nil {
+		e := err.(*googleapi.Error)
+		if e.Code == 404 {
+			// Don't treat not found deployment deletion as error to make kfctl delete idempotent.
+			log.Infof("Deployment %v/%v is not found during deletion.", project, name)
+			return nil
+		} else {
+			return fmt.Errorf("Deployment %v/%v has unexpected error: %v", project, name, err)
+		}
+	}
+
+	op, err := deploymentmanagerService.Deployments.Delete(project, name).Context(ctx).Do()
+	if err != nil {
+		return fmt.Errorf("Gcp.Delete is failed for %v/%v: %v", project, name, err)
+	}
+	if err = blockingWait(project, op.Name, deploymentmanagerService, ctx,
+		"Deleting "+name); err != nil {
+		return fmt.Errorf("Gcp.Delete is failed for %v/%v: %v", project, name, err)
+	}
+	return nil
+}
+
 func (gcp *Gcp) Delete(resources kftypes.ResourceEnum) error {
+	ctx := context.Background()
+	client, err := google.DefaultClient(ctx, deploymentmanager.CloudPlatformScope)
+	if err != nil {
+		return fmt.Errorf("Error getting DefaultClient: %v", err)
+	}
+	deploymentmanagerService, err := deploymentmanager.New(client)
+	if err != nil {
+		return fmt.Errorf("Error creating deploymentmanagerService: %v", err)
+	}
+
+	// cluster and storage deployments are required to be deleted. network and gcfs deployments are optional.
+	project := gcp.Spec.Project
+	deletingDeployments := []string{
+		gcp.Name,
+	}
+	if gcp.Spec.DeleteStorage {
+		deletingDeployments = append(deletingDeployments, gcp.Name+"-storage")
+	}
+	if _, networkStatErr := os.Stat(path.Join(gcp.Spec.AppDir, NETWORK_FILE)); !os.IsNotExist(networkStatErr) {
+		deletingDeployments = append(deletingDeployments, gcp.Name+"-network")
+	}
+	if _, gcfsStatErr := os.Stat(path.Join(gcp.Spec.AppDir, GCFS_FILE)); !os.IsNotExist(gcfsStatErr) {
+		deletingDeployments = append(deletingDeployments, gcp.Name+"-gcfs")
+	}
+
+	for _, d := range deletingDeployments {
+		if err = deleteDeployment(deploymentmanagerService, ctx, project, d); err != nil {
+			return err
+		}
+	}
+
+	policy, err := utils.GetIamPolicy(project)
+	if err != nil {
```

Comment:
should we retry?

Reply:
during my local testing this part is not unstable that I think it needs retry.

Reply:
not unstable --> no need retry?
or unstable --> need retry?

Reply:
not unstable*